options/posix/generic/posix-file-io.cpp

```diff
@@ -6,12 +6,39 @@
 
 namespace mlibc {
 
-mem_file::mem_file(char **ptr, size_t *sizeloc, void (*do_dispose)(abstract_file *))
-: abstract_file{do_dispose}, _bufloc{ptr}, _sizeloc{sizeloc}, _buf{getAllocator()}, _pos{0} { }
+// constructor for `open_memstream`-style
+mem_file::mem_file(char **ptr, size_t *sizeloc, int flags, void (*do_dispose)(abstract_file *), bool allow_resizing)
+: abstract_file{do_dispose}, _bufloc{ptr}, _sizeloc{sizeloc}, _flags{flags},
+	_allowResizing{allow_resizing} { }
+
+// constructor for `fmemopen`-style
+mem_file::mem_file(void *in_buf, size_t size, int flags, void (*do_dispose)(abstract_file *))
+: abstract_file{do_dispose}, _inBuffer{in_buf}, _inBufferSize{size}, _flags{flags},
+	_allowResizing{false} {
+	if(!_inBuffer) {
+		_inBuffer = getAllocator().allocate(size);
+		_needsDeallocation = true;
+	}
+
+	if(_flags & O_APPEND) {
+		// the initial seek-size for append is zero if buf was NULL, or the first '\0' found, or the size
+		_max_size = (_needsDeallocation) ? 0 : strnlen(reinterpret_cast<char *>(_inBuffer), _inBufferSize);
```

Comment:
While this is correct, there is missing functionality here, which is that append should also set the position of the stream to the first null byte. This functionality is also not tested in the tests afaict, which is why that hasn't been caught.